aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -1096,6 +1096,128 @@ Tensor clone(const Tensor& src, c10::optional<c10::MemoryFormat> optional_memory
+  const Tensor& aweights = c10::value_or_else(aweightsOpt, [] {return Tensor();});
+  auto bias = biasOpt.value_or(false);
+  auto ddof = ddofOpt.value_or(0);
+  auto rowvar = rowvarOpt.value_or(true);
+
+  if(self.size(0) == 0)
+    return at::empty_like(self);  
+  TORCH_CHECK(self.dim() <= 2,
+              "Input1 has more than 2 dimensions." );
+    
+  Tensor X, y;
+  if(self.dim() == 1)
+    X = self.unsqueeze(0);
+  else
+    X = self;
+
+  if(!rowvar && X.size(0) != 1)
+    X = X.t();
+
+  if(other.defined()) {
+    TORCH_CHECK(other.dim() <= 2,
+                "Input2 has more than 2 dimensions." );
+ 
+    if(other.dim() == 1)
+      y = other.unsqueeze(0);
+    else
+      y = other;
+
+    if(!rowvar && y.size(0) != 1)
+      y = y.t();
+
+    std::vector<Tensor> t_list;
+    t_list.push_back(X); t_list.push_back(y);
+    X = at::cat(t_list, 0);
+  }
+
+  if (ddofOpt == c10::nullopt) {
+    ddof = (bias == false) ? 1 : 0;
+  }
+  
+  Tensor w = at::empty(0,self.options());
+  
+  if(fweights.defined()) {
+    TORCH_CHECK((fweights.scalar_type() == at::kLong || fweights.scalar_type() == at::kInt),
+                "fweights must be integer.");
+    TORCH_CHECK(fweights.dim() <= 1,
+                "cannot handle multidimensional fweights.");
+    TORCH_CHECK(fweights.size(0) == X.size(1),
+                "incompatible numbers of samples and weights.");
+ 
+    for (auto i =0; i<fweights.size(0); i++){
+      if (fweights[i].item<int>() < 0)
+        TORCH_CHECK(false,
+                  "fweights cannot be negative.");
+    }
+    w = at::clone(fweights);
+  }
+
+  if (aweights.defined()) {
+    TORCH_CHECK(aweights.dim() <= 1,
+                "cannot handle multidimensional aweights.");
+    TORCH_CHECK(aweights.size(0) == X.size(1),
+                "incompatible numbers of samples and weights.");
+    for (auto i =0; i<aweights.size(0); i++){
+      if (aweights[i].item<float>() < 0)
+        TORCH_CHECK(false,
+                  "aweights cannot be negative.");
+    }
+
+    if(w.size(0) == 0)
+      w = at::clone(aweights);
+    else
+      w *= aweights;
+  }
+
+  w = w.to(at::kFloat);
+  Tensor avg, w_sum;
+  if(w.size(0) != 0) {
+    avg = at::sum(X.mul(w), 1) / (at::sum(w));
+    w_sum = at::sum(w);
+  }
+  else {
+    avg = at::mean(X,1);
+    w_sum = at::ones(X.size(0))*(X.size(1));
+  }
+
+  Tensor fact = at::empty({0}, self.options()); 
+  if(w.size(0) == 0) {
+    fact = at::tensor(X.size(1) - ddof, self.options());
+  }
+  else if(ddof == 0)
+    fact = w_sum;
+  else if(aweights.size(0) == 0)
+    fact = w_sum - ddof;
+  else
+    fact = w_sum - ddof*at::sum(w*aweights)/w_sum; 
+  
+  if(fact.item<int>() <= 0) {
+    fact = at::zeros({1}, self.options());
+  } 
+
+  X = X - avg.unsqueeze(1);
+  Tensor X_t = (w.size(0) == 0) ? X.t() : (X*w).t();
+  Tensor c = at::mm(X,X_t);
+  c = at::true_divide(c, fact);
+
+  return c.squeeze().to(self.dtype());
```

Comment:
it's concerning that you need this `to(self.dtype()` cast here - what are the situations where c naturally won't be of the correct type? I think they should be handled right away, so that `c` is automatically produced with the correct dtype

Reply:
My bad, `c` indeed produces the correct dtype by itself. Thanks for pointing it out!